File: java-player/src/main/java/ubc/cosc322/messages/MessageParser.java
package ubc.cosc322.messages;

import com.google.gson.Gson;
import com.google.gson.JsonParser;

public class MessageParser {
    public Gson gson;
    public JsonParser parser = new JsonParser();

    public MessageParser(Gson gson) {
        this.gson = gson;
    }

    public BaseMessage parse(String msg)
    {
        BaseMessage parsed = gson.fromJson(msg, BaseMessage.class);
        switch (parsed.message_type) {
          case MessageType.STATE_UPDATE:
            return gson.fromJson(msg, StateUpdate.class);
          case MessageType.ROLE_NEGOTIATION:
            return gson.fromJson(msg, RoleNegotiation.class);
          case MessageType.ROOMLIST_MESSAGE:
            return gson.fromJson(msg, RoomListMessage.class);
          case MessageType.LOGGING_MESSAGE:
            return gson.fromJson(msg, LoggingMessage.class);
          case MessageType.ROOM_CHANGE_MESSAGE:
            return gson.fromJson(msg, RoomChangeMessage.class);
          default:
            return null;
        }
    }
}

File: shared/messages/message-parser.ts
import WebSocket from "ws";
import { BaseMessage } from "./base-message";
import { MessageType } from "../enum/message-type";
import { StateUpdate } from "./state-update";
import { RoleNegotiation } from "./role-negotiation";
import { RoomMessage } from "./room-message";
import { PlayerListMessage } from "./playerlist-message";
import { LoggingMessage } from "./logging-message";
import { RoomChangeMessage } from "./room-change-message";
import { MCTSUpdateMessage } from "./mcts-update-message";

export class MessageParser {
  parse(msg: WebSocket.Data): BaseMessage | undefined {
    const message: string = msg.toString();
    var parsed = JSON.parse(message);

    switch (parsed.message_type) {
      case MessageType.STATE_UPDATE:
        return new StateUpdate(parsed);
      case MessageType.ROLE_NEGOTIATION:
        return new RoleNegotiation(parsed);
      case MessageType.ROOMLIST_MESSAGE:
        return new RoomMessage(parsed);
      case MessageType.PLAYERLIST_MESSAGE:
        return new PlayerListMessage(parsed);
      case MessageType.LOGGING_MESSAGE:
        return new LoggingMessage(parsed);
      case MessageType.ROOM_CHANGE_MESSAGE:
        return new RoomChangeMessage(parsed);
      case MessageType.MCTS_UPDATE_MESSAGE:
        return new MCTSUpdateMessage(parsed);
    }

    return undefined;
  }
}
